hotel/models/hotel_folio_room.py

```diff
@@ -0,0 +1,234 @@
+from openerp.osv import fields
+import time
+from openerp import netsvc
+from openerp import models
+from openerp import fields
+from openerp import api
+from mx import DateTime
+import datetime
+from openerp.tools import DEFAULT_SERVER_DATETIME_FORMAT
+from openerp.tools import config
+from openerp.exceptions import except_orm
+from openerp.exceptions import Warning
+from openerp.exceptions import RedirectWarning
+from openerp.exceptions import ValidationError
+from openerp.tools.translate import _
+
+
+class HotelFolioRoom(models.Model):
+    _name = 'hotel.folio.room'
+    _description = 'Hotel Folio Room'
+
+    @api.one
+    def copy(self, default=None):
+        copy = self.env['sale.order.line'].browse(self.id)
+        return copy.copy(default=None)
+
+    @api.multi
+    def _amount_line_net(self, field_name, arg):
+        amount_line = self.env['sale.order.line'].browse(self.ids)
+        return amount_line._amount_line_net(field_name, arg)
+
+    @api.multi
+    def _amount_line(self, field_name, arg):
+        amount_line = self.env['sale.order.line'].browse(self.ids)
+        return amount_line._amount_line(field_name, arg)
+
+    @api.multi
+    def _number_packages(self, field_name, arg):
+        packages = self.env['sale.order.line'].browse(self.ids)
+        return packages._number_packages(field_name, arg)
+
+    @api.model
+    def _get_1st_packaging(self):
+        return self.env['sale.order.line']._get_1st_packaging()
+
+    @api.model
+    def _get_checkin_date(self):
+        if 'checkin_date' in self._context:
+            return self._context['checkin_date']
+        return time.strftime('%Y-%m-%d %H:%M:%S')
+
+    @api.model
+    def _get_checkout_date(self):
+        if 'checkin_date' in self._context:
+            return self._context['checkout_date']
+        return time.strftime('%Y-%m-%d %H:%M:%S')
+
+    order_line_id = fields.Many2one(
+        'sale.order.line',
+        string='Order Line',
+        required=True,
+        ondelete='cascade',
+        delegate=True)
+    folio_id = fields.Many2one(
+        'hotel.folio',
+        string='Folio',
+        ondelete='cascade',
+        required=True)
+    checkin_date = fields.Datetime(
+        string='Check In',
+        required=True)
+    checkout_date = fields.Datetime(
+        string='Check Out',
+        required=True)
+
+    _defaults = {
+       'checkin_date': _get_checkin_date,
+       'checkout_date': _get_checkout_date,
+    }
+
+    @api.model
+    def create(self, vals, check=True):
+        folio = self.env['hotel.folio'].browse([vals['folio_id']])[0]
+        vals.update({'order_id': folio.order_id.id})
+        return super(HotelFolioRoom, self).create(vals)
+
+    @api.onchange('product_id')
+    @api.depends('product_id')
```

Comment:
I am not sure that the @api.depends is needed if you already have the @api.onchange on that field. Especially since you start this method by:

    if self.product_id: